src/display/unicode.c

```diff
@@ -0,0 +1,102 @@
+   see Wikipedia note. https://bit.ly/39t1732 */
+
+wchar_t CP437_to_unicode[256] = {
+    0x20, 0x263a, 0x263b, 0x2665, 0x2666, 0x2663, 0x2660, 0x2022,
+    0x25d8, 0x25cb, 0x25d9, 0x2642, 0x2640, 0x266a, 0x266b, 0x263c,
+    0x25ba, 0x25c4, 0x2195, 0x203c, 0xb6, 0xa7, 0x25ac, 0x21a8,
+    0x2191, 0x2193, 0x2192, 0x2190, 0x221f, 0x2194, 0x25b2, 0x25bc,
+
+    0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27,
+    0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
+    0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+    0x38, 0x39, 0x3a, 0x3b, 0x3c, 0x3d, 0x3e, 0x3f,
+    0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47,
+    0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f,
+    0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57,
+    0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
+    0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67,
+    0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f,
+    0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
+    0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x2302,
+
+    0xc7, 0xfc, 0xe9, 0xe2, 0xe4, 0xe0, 0xe5, 0xe7,
+    0xea, 0xeb, 0xe8, 0xef, 0xee, 0xec, 0xc4, 0xc5,
+    0xc9, 0xe6, 0xc6, 0xf4, 0xf6, 0xf2, 0xfb, 0xf9,
+    0xff, 0xd6, 0xdc, 0xa2, 0xa3, 0xa5, 0x20a7, 0x192,
+    0xe1, 0xed, 0xf3, 0xfa, 0xf1, 0xd1, 0xaa, 0xba,
+    0xbf, 0x2310, 0xac, 0xbd, 0xbc, 0xa1, 0xab, 0xbb,
+    0x2591, 0x2592, 0x2593, 0x2502, 0x2524, 0x2561, 0x2562, 0x2556,
+    0x2555, 0x2563, 0x2551, 0x2557, 0x255d, 0x255c, 0x255b, 0x2510,
+    0x2514, 0x2534, 0x252c, 0x251c, 0x2500, 0x253c, 0x255e, 0x255f,
+    0x255a, 0x2554, 0x2569, 0x2566, 0x2560, 0x2550, 0x256c, 0x2567,
+    0x2568, 0x2564, 0x2565, 0x2559, 0x2558, 0x2552, 0x2553, 0x256b,
+    0x256a, 0x2518, 0x250c, 0x2588, 0x2584, 0x258c, 0x2590, 0x2580,
+    0x3b1, 0xdf, 0x393, 0x3c0, 0x3a3, 0x3c3, 0xb5, 0x3c4,
+    0x3a6, 0x398, 0x3a9, 0x3b4, 0x221e, 0x3d5, 0x3b5, 0x2229,
+    0x2261, 0xb1, 0x2265, 0x2264, 0x2320, 0x2321, 0xf7, 0x2248,
+    0xb0, 0x2219, 0xb7, 0x221a, 0x207f, 0xb2, 0x25a0, 0xa0
+};
+
+void init_unicode_conversion() {
+	setlocale(LC_CTYPE, "en_US.utf8");
+	bzero(unicode_to_CP437, 65536);
+
+	/* First reverse the cp437 to unicode codepoint table. */
+	for (int i = 0; i <= 255; ++i) {
+		unicode_to_CP437[CP437_to_unicode[i]] = i;
+	}
+
+	/* Then add some "near-equivalents" */
+	/*lowercase ø: make it look like phi*/
+	unicode_to_CP437[248] = 237;
+
+	/*uppercase Ø: make it look like Theta*/
+	unicode_to_CP437[216] = 233;
+
+	unicode_table_inited = 1;
+}
+
+char get_CP437_from_UTF8(char * utf8_str) {
+
+	mbstate_t state = {0};
+	wchar_t codepoint;
+
+	/* Alternatively do if (!unicode_conversion_inited) init_unicode_conversion();*/
+	assert(unicode_table_inited);
+
+	size_t bytes_converted = mbrtowc(&codepoint, utf8_str,
+		min(2, strlen(utf8_str)), &state);
+
+	/* Check if something went wrong. If so, abort. */
+	if (bytes_converted == (size_t)-1 || bytes_converted == (size_t)-2) {
+		return 0;
+	}
+
+	if (codepoint >= 65536 || unicode_to_CP437[codepoint] == 0) {
```

Comment:
The last condition can be dropped, since 0 is what gets returned at the end of this function if `unicode_to_CP437[codepoint]` is 0.

```suggestion
	if (codepoint >= 65536) {
```

Reply:
Fixed.